Parse TLE to KeplerInputs

In [37]:
import sys
sys.path.append('/workspaces/ma-max-wendler/scripts/keplertraces')  # Add the directory to the search path

In [38]:
import tleparse

In [39]:
# inputs = tleparse($tlespath$), i.e.
path = "/workspaces/ma-max-wendler/examples/space_veins/tles/iridiumNEXT_2023-10-22-21-39-19.txt"
tles = tleparse.read(path)

In [40]:
# objects of the TLE-tool that among other things parses TLE values
tle = tles[0]
tle

TLE(name='IRIDIUM 106', norad='41917', classification='U', int_desig='17003A', epoch_year=2023, epoch_day=295.91774854, dn_o2=1.91e-06, ddn_o6=0.0, bstar=6.0989e-05, set_num=999, inc=86.3949, raan=133.2138, ecc=0.0001982, argp=93.4819, M=266.6603, n=14.34216513, rev_num=35442)

For comparison: IRIDIUM 106 TLE

```
IRIDIUM 106
1 41917U 17003A   23295.91774854  .00000191  00000+0  60989-4 0  9995
2 41917  86.3949 133.2138 0001982  93.4819 266.6603 14.34216513354425
```

Convert to KeplerInputs instance, contains two conversions:

In [41]:
from poliastro.twobody.angles import M_to_E, E_to_nu
import astropy.units as u
from math import pi

In [42]:
# mean anomaly to true anomaly via poliastro functions
ecc = u.Quantity(tle.ecc, unit=u.one)
mean_anom = u.Quantity(tle.M, unit=u.deg)
ecc, mean_anom

(<Quantity 0.0001982>, <Quantity 266.6603 deg>)

In [43]:
mean_anom_rad = mean_anom.to(u.rad)
# mean_anom_rad_negpos = u.Quantity(pi,unit=u.rad) - mean_anom.to(u.rad)
# -pi (when pi passed) + (mean_anom_rad - pi) (overlap over pi)
mean_anom_rad_negpos = mean_anom.to(u.rad) - u.Quantity(2*pi,unit=u.rad) 

mean_anom_rad, mean_anom_rad_negpos, 2*pi + mean_anom_rad_negpos.value

(<Quantity 4.65410022 rad>, <Quantity -1.62908509 rad>, 4.654100219355835)

In [44]:
import tletools

In [45]:
"""
**Warning**

    The mean anomaly must be between -π and π radians.
    The eccentricity must be less than 1.
"""
    
# tle-tools-based true anomaly
true_anom_t = u.Quantity(tletools.utils.M_to_nu(mean_anom_rad_negpos, tle.ecc), unit=u.rad).to(u.deg)
true_anom_t, true_anom_t.to(u.rad)

(<Quantity -93.36237315 deg>, <Quantity -1.62948081 rad>)

In [46]:
# poliastro based True anomaly -> can use mean anomaly in degrees
true_anom_p = E_to_nu( M_to_E( mean_anom_rad_negpos, ecc ), ecc )
_true_anom_p = E_to_nu( M_to_E( mean_anom_rad, ecc ), ecc )
true_anom_p, _true_anom_p, true_anom_p.to(u.deg)

(<Quantity -1.62948081 rad>,
 <Quantity -1.62948081 rad>,
 <Quantity -93.36237315 deg>)

In [47]:
from orbital.utilities import true_anomaly_from_mean

In [48]:
# OrbitalPy based True anomaly
# uses radian in 0,2pi range (Mnorm = fmod(M, 2 * pi))
true_anom_o = true_anomaly_from_mean(ecc.value, mean_anom_rad.value)
true_anom_o, u.Quantity( -1 * (2*pi - true_anom_o), unit=u.rad).to(u.deg)

(4.653704498287206, <Quantity -93.36237315 deg>)

In [49]:
import rebound

In [50]:
# rebound-based True anomaly
# can't really tell in which range M should be, as its C code
true_anom_r = rebound.M_to_f(tle.ecc, mean_anom_rad_negpos.value)
true_anom_r, u.Quantity(-360, unit=u.deg) + u.Quantity(true_anom_r, unit=u.rad).to(u.deg)

(4.653704498287205, <Quantity -93.36237315 deg>)

In [51]:
from skyfield import keplerlib

In [52]:
# can take mean anomalies in both ranges
ecc_anom = keplerlib.eccentric_anomaly(tle.ecc, mean_anom_rad.value)
_ecc_anom = keplerlib.eccentric_anomaly(tle.ecc, mean_anom_rad_negpos.value)
true_anom_s = keplerlib.true_anomaly_closed(tle.ecc, ecc_anom)
_true_anom_s = keplerlib.true_anomaly_closed(tle.ecc, _ecc_anom)

true_anom_s, u.Quantity(true_anom_s, unit=u.rad).to(u.deg), _true_anom_s

(-1.629480808892381, <Quantity -93.36237315 deg>, -1.629480808892381)

Semi-major axis

In [53]:
# tletools 
tle.a

7155.808268901222

In [54]:
from astropy.constants import GM_earth

In [55]:
# researched formula
secs_per_revolution = 86400 / tle.n
semimajoraxis = u.Quantity((secs_per_revolution**2 * GM_earth.value / (4 * pi**2) )**(1/3) / 1000, unit=u.km)
semimajoraxis

<Quantity 7155.80801877 km>

TEME coordinates from poliastro orbit

In [56]:
# creates kepler orbit
orb = tle.to_orbit()
orb

7154 x 7157 km x 86.4 deg (GCRS) orbit around Earth (♁) at epoch 2023-10-22T22:01:33.473856000 (UTC)

Show that equivalent to reference calculations from literature / validated above:

In [57]:
from keplerinputs import KeplerInputs
from poliastro.twobody import Orbit
from poliastro.bodies import Earth

In [58]:
kepler_inputs = KeplerInputs(tle.name, tle.ecc, tle.inc, tle.raan, tle.argp, mean_anom=tle.M, mean_motion=tle.n, epoch=tle.epoch)
orb = Orbit.from_classical(
            Earth, 
            a=kepler_inputs.semimajoraxis, 
            ecc=kepler_inputs.ecc, 
            inc=kepler_inputs.inc, 
            raan=kepler_inputs.raan, 
            argp=kepler_inputs.argp, 
            nu=kepler_inputs.true_anom, 
            epoch=kepler_inputs.epoch)
orb

7154 x 7157 km x 86.4 deg (GCRS) orbit around Earth (♁) at epoch 2023-10-22T22:01:33.473856000 (UTC)

=> tletools orbit creation is equivalent.

Now: compare ephem coordinates to TEME coordinates from C++ SGP4

In [59]:
from datetime import datetime
from astropy.time import Time

In [60]:
# following avg. epoch of all sats, which is encoded in name of TLEs file, 'iridium-NEXT_2023-10-22-21-39-19'
start_datetime = datetime.strptime("2023-10-22-21-39-19", '%Y-%m-%d-%H-%M-%S')
sim_start_time = Time(start_datetime, format='datetime', scale='utc')
sim_start_time

<Time object: scale='utc' format='datetime' value=2023-10-22 21:39:19>

In [61]:
from poliastro.twobody.sampling import EpochsArray
from poliastro.util import time_range

In [62]:
# could use different propagator for EpochsArray, which currently is 'FarnocchiaPropagator'
update_interval = (1 << u.s)
periods = 25
# first coordinate change at first sim second
ephem = orb.to_ephem(strategy=EpochsArray(epochs=time_range(sim_start_time + update_interval, spacing=update_interval, periods=periods)))
ephem

Ephemerides at 25 epochs from 2023-10-22 21:39:20 (UTC) to 2023-10-22 21:39:44 (UTC)

In [63]:
# ICRS coordinates, but orbit above specifies GCRS?
# => GCRS is specific version of ICRS
"""
Regardless of how the Orbit is created, 
the implicit reference system is an inertial one. 
For the specific case of the Solar System, this can be assumed 
to be the International Celestial Reference System or ICRS.
"""

cartesian_trace = ephem.sample(ephem.epochs)
# example coordinates
cartesian_trace

<CartesianRepresentation (x, y, z) in km
    [(-566.52604985, 1249.35180412, -7024.52179864),
     (-571.61236662, 1254.64088416, -7023.16674788),
     (-576.69806194, 1259.92860018, -7021.80406161),
     (-581.78313028, 1265.21494641, -7020.43374131),
     (-586.86756611, 1270.49991712, -7019.05578846),
     (-591.9513639 , 1275.78350655, -7017.67020456),
     (-597.03451813, 1281.06570896, -7016.27699111),
     (-602.11702327, 1286.3465186 , -7014.87614964),
     (-607.1988738 , 1291.62592975, -7013.46768166),
     (-612.28006418, 1296.90393665, -7012.05158871),
     (-617.3605889 , 1302.18053357, -7010.62787232),
     (-622.44044243, 1307.45571477, -7009.19653404),
     (-627.51961925, 1312.72947451, -7007.75757543),
     (-632.59811383, 1318.00180707, -7006.31099804),
     (-637.67592066, 1323.27270671, -7004.85680346),
     (-642.75303422, 1328.5421677 , -7003.39499326),
     (-647.82944897, 1333.81018431, -7001.92556903),
     (-652.90515942, 1339.07675081, -7000.44853236),
     

In [64]:
from astropy.coordinates import GCRS, ITRS, WGS84GeodeticRepresentation, SkyCoord

In [65]:
# validated: is GCRS
gcrs_trace = GCRS(cartesian_trace, obstime=ephem.epochs)
gcrs_trace.cartesian[:3]

<CartesianRepresentation (x, y, z) in km
    [(-566.52604985, 1249.35180412, -7024.52179864),
     (-571.61236662, 1254.64088416, -7023.16674788),
     (-576.69806194, 1259.92860018, -7021.80406161)]
 (has differentials w.r.t.: 's')>

In [66]:
# convert GCRS to ITRF/S
itrf_trace = gcrs_trace.transform_to(ITRS(obstime=ephem.epochs))
itrf_trace.cartesian[:3]

<CartesianRepresentation (x, y, z) in km
    [(-644.86301954, 1203.58311841, -7025.75814829),
     (-650.25598513, 1208.51251027, -7024.41458634),
     (-655.64752835, 1213.44138109, -7023.06338752)]
 (has differentials w.r.t.: 's')>

In [69]:
# transform to SkyCoord
skycoord_itrf = SkyCoord(itrf_trace, obstime=ephem.epochs, frame="itrs")
skycoord_itrf.cartesian[15:]

<CartesianRepresentation (x, y, z) in km
    [(-725.60553853, 1277.46676615, -7004.80352685),
     (-730.97656521, 1282.3877606 , -7003.34556976),
     (-736.34608341, 1287.30815087, -7001.8799987 ),
     (-741.71408742, 1292.22793144, -7000.40681525),
     (-747.08057153, 1297.14709674, -6998.92602102),
     (-752.44553001, 1302.06564125, -6997.43761762),
     (-757.80895715, 1306.98355941, -6995.94160666),
     (-763.17084724, 1311.90084569, -6994.43798978),
     (-768.53119458, 1316.81749452, -6992.92676859),
     (-773.88999345, 1321.73350039, -6991.40794475)]
 (has differentials w.r.t.: 's')>

For comparison, output by orbkit transformation:

```
[ <PVCoordinates: {P(-727.0938966215228, 1276.6212194343068, -7004.802506545653), V(0.09309187026826966, 0.05301996231850821, -5.3028588578756396E-8), A(3.866278039020582E-6, -6.788368278032022E-6, -1.4665325824003472E-11)}>,
 <PVCoordinates: {P(-732.5641094702186, 1281.4825284829526, -7003.344549410315), V(0.09344636266366925, 0.05341885667975937, -5.296804106243908E-8), A(3.89536587490266E-6, -6.814218271316078E-6, -1.4740768365942567E-11)}>,
 <PVCoordinates: {P(-738.0335258941184, 1286.342444312725, -7001.878978291198), V(0.09380075346468401, 0.05381769296553489, -5.290743596137211E-8), A(3.924449475852732E-6, -6.840060856218541E-6, -1.4816194881918557E-11)}>,
 <PVCoordinates: {P(-743.5021399256423, 1291.2009616473915, -7000.405794805787), V(0.09415504228656484, 0.0542164707406718, -5.2846773344231464E-8), A(3.953528810138209E-6, -6.8658960046830585E-6, -1.4891605289828644E-11)}>,
 <PVCoordinates: {P(-748.9699456264644, 1296.0580751984296, -6998.925000521637), V(0.09450922874366643, 0.05461518957214039, -5.278605327566034E-8), A(3.982603846182069E-6, -6.891723688587936E-6, -1.496699950769187E-11)}>,
 <PVCoordinates: {P(-754.4369370496694, 1300.913779698044, -6997.436597076279), V(0.09486331245185493, 0.05501384902628459, -5.272527582314434E-8), A(4.011674552361613E-6, -6.917543879921688E-6, -1.504237745361741E-11)}>,
 <PVCoordinates: {P(-759.9031082560932, 1305.768069847746, -6995.940586067266), V(0.09521729302475829, 0.055412448670013555, -5.2664441050442644E-8), A(4.040740897095359E-6, -6.943356550509619E-6, -1.5117739045544903E-11)}>,
 <PVCoordinates: {P(-765.3684533087905, 1310.6209403789694, -6994.436969132149), V(0.09557117007818644, 0.05581098807039825, -5.260354902399753E-8), A(4.069802848813619E-6, -6.96916167233615E-6, -1.5193084201650194E-11)}>,
 <PVCoordinates: {P(-770.8329662409272, 1315.4723860254567, -6992.9257478984155), V(0.09592494322811763, 0.05620946679233008, -5.25425998133788E-8), A(4.098860375787778E-6, -6.994959217397972E-6, -1.5268412839909278E-11)}>,
 <PVCoordinates: {P(-776.2966411431365, 1320.322401486421, -6991.406924013685), V(0.09627861208801226, 0.05660788440489108, -5.2481593479228013E-8), A(4.127913446594798E-6, -7.020749157508173E-6, -1.5343724878461044E-11)}>]
```

For comparison, output by C++ SGP4
```
> print(x_sgp4_itrf$`itrfCoordX_vector`)
 [1] -734.9439 -740.2763 -745.6072 -750.9365 -756.2643 -761.5906 -766.9153 -772.2386 -777.5602 -782.8803 -788.1988
> print(y_sgp4_itrf$`itrfCoordY_vector`)
 [1] 1280.705 1285.646 1290.587 1295.527 1300.466 1305.405 1310.343 1315.280 1320.217 1325.153 1330.089
> print(z_sgp4_itrf$`itrfCoordZ_vector`)
 [1] -7010.836 -7009.370 -7007.896 -7006.415 -7004.926 -7003.430 -7001.926 -7000.414 -6998.895 -6997.368 -6995.834
```

In [32]:
# transform to WGS84
skycoord_wgs84 = skycoord_itrf.copy()
skycoord_wgs84.set_representation_cls(WGS84GeodeticRepresentation)
skycoord_wgs84[15:]

<SkyCoord (ITRS: obstime=[datetime.datetime(2023, 10, 22, 21, 39, 35)
 datetime.datetime(2023, 10, 22, 21, 39, 36)
 datetime.datetime(2023, 10, 22, 21, 39, 37)
 datetime.datetime(2023, 10, 22, 21, 39, 38)
 datetime.datetime(2023, 10, 22, 21, 39, 39)
 datetime.datetime(2023, 10, 22, 21, 39, 40)
 datetime.datetime(2023, 10, 22, 21, 39, 41)
 datetime.datetime(2023, 10, 22, 21, 39, 42)
 datetime.datetime(2023, 10, 22, 21, 39, 43)
 datetime.datetime(2023, 10, 22, 21, 39, 44)], location=(0., 0., 0.) km): (lon, lat, height) in (deg, deg, m)
    [(119.59672477, -78.22323616, 799563.2714196 ),
     (119.68366428, -78.16682804, 799554.57686177),
     (119.7697517 , -78.11039084, 799545.8381859 ),
     (119.85499866, -78.05392496, 799537.05542884),
     (119.93941662, -77.99743081, 799528.22862769),
     (120.02301682, -77.94090878, 799519.35781967),
     (120.10581031, -77.88435925, 799510.44304222),
     (120.18780795, -77.82778259, 799501.48433297),
     (120.26902042, -77.77117919, 799492.481

For comparison, outputs by C++ SGP4:

```
> print(x_sgp4$`wgs84CoordLat_vector`)
 [1] -78.17508 -78.11866 -78.06220 -78.00573 -77.94922 -77.89269 -77.83613 -77.77955 -77.72294 -77.66631 -77.60965
> print(y_sgp4$`wgs84CoordLon_vector`)
 [1] 119.8497 119.9334 120.0162 120.0982 120.1794 120.2599 120.3395 120.4184 120.4966 120.5739 120.6506
> print(z_sgp4$`wgs84CoordAlt_vector`)
 [1] 806989.9 806980.9 806971.8 806962.7 806953.5 806944.3 806935.0 806925.7 806916.4 806907.0 806897.6

```

In [33]:
# even larger mismatch in altitude
skycoord_wgs84.spherical[15:] 

<SphericalRepresentation (lon, lat, distance) in (deg, deg, km)
    [(119.59672477, -78.1547217 , 7157.2129482 ),
     (119.68366428, -78.09800433, 7157.21274498),
     (119.7697517 , -78.04125799, 7157.21254023),
     (119.85499866, -77.98448309, 7157.21233395),
     (119.93941662, -77.92768003, 7157.21212615),
     (120.02301682, -77.8708492 , 7157.21191682),
     (120.10581031, -77.813991  , 7157.21170596),
     (120.18780795, -77.75710581, 7157.21149358),
     (120.26902042, -77.700194  , 7157.21127967),
     (120.34945822, -77.64325594, 7157.21106423)]
 (has differentials w.r.t.: 's')>

In [34]:
# mismatch in angles
data = skycoord_wgs84.spherical[15:].represent_as(WGS84GeodeticRepresentation)
[(d.lon.to(u.deg), d.lat.to(u.deg), d.height) for d in data]

[(<Longitude 119.59672477 deg>,
  <Latitude -78.22323616 deg>,
  <Quantity 799563.2714196 m>),
 (<Longitude 119.68366428 deg>,
  <Latitude -78.16682804 deg>,
  <Quantity 799554.57686178 m>),
 (<Longitude 119.7697517 deg>,
  <Latitude -78.11039084 deg>,
  <Quantity 799545.8381859 m>),
 (<Longitude 119.85499866 deg>,
  <Latitude -78.05392496 deg>,
  <Quantity 799537.05542884 m>),
 (<Longitude 119.93941662 deg>,
  <Latitude -77.99743081 deg>,
  <Quantity 799528.22862769 m>),
 (<Longitude 120.02301682 deg>,
  <Latitude -77.94090878 deg>,
  <Quantity 799519.35781967 m>),
 (<Longitude 120.10581031 deg>,
  <Latitude -77.88435925 deg>,
  <Quantity 799510.44304222 m>),
 (<Longitude 120.18780795 deg>,
  <Latitude -77.82778259 deg>,
  <Quantity 799501.48433297 m>),
 (<Longitude 120.26902042 deg>,
  <Latitude -77.77117919 deg>,
  <Quantity 799492.48172972 m>),
 (<Longitude 120.34945822 deg>,
  <Latitude -77.7145494 deg>,
  <Quantity 799483.43527044 m>)]

In [35]:
# not equivalent -> ITRS conversion is necessara
data = SkyCoord(GCRS(cartesian_trace, obstime=ephem.epochs), frame="gcrs", obstime=ephem.epochs)
data = data.spherical.represent_as(WGS84GeodeticRepresentation)[15:]
[(d.lon * 180 / pi) for d in data]

[<Angle 115.81784858 rad>,
 <Angle 115.90577074 rad>,
 <Angle 115.99287188 rad>,
 <Angle 116.07916326 rad>,
 <Angle 116.16465589 rad>,
 <Angle 116.24936064 rad>,
 <Angle 116.33328815 rad>,
 <Angle 116.4164489 rad>,
 <Angle 116.49885319 rad>,
 <Angle 116.58051113 rad>]